File: src/actions_py/actions_py/robot_move_client.py
import rclpy, time
from rclpy.node import Node
from my_robot_interfaces.action import RobotMove
from rclpy.action import ActionClient
from action_msgs.msg import GoalStatus
from std_msgs.msg import Empty

class RobotMoveActionClient(Node):

    def __init__(self):
        super().__init__('robot_move_client')
        self._action_client = ActionClient(self, RobotMove, 'robot_move')
        self._sent_goals = 0
        self._received_results = 0
        self._goal_handles = {}

        # Create a subscriber to the Empty message for cancelling any robot move
        self._cancel_robot_move_subscriber = self.create_subscription(
            Empty,
            'cancel_robot_move',
            self.cancel_robot_move,
            10
        )

    def send_goal(self, position, velocity, cancel=False):
        goal_msg = RobotMove.Goal()
        goal_msg.position = position
        goal_msg.velocity = velocity

        self._action_client.wait_for_server()
        goal_id = self._sent_goals
        self._sent_goals += 1
        self._send_goal_future = self._action_client.send_goal_async(goal_msg, feedback_callback=self.feedback_callback)
        self._send_goal_future.add_done_callback(lambda future: self.goal_response_callback(future, goal_id))

        # Send a cancel request after 2 seconds
        if cancel:
            self._timer = self.create_timer(2.0, lambda: self.send_cancel_request(goal_id, cancel_timer=True))
    

    def cancel_robot_move(self, msg):
        for goal_id in self._goal_handles:
            if not self._goal_handles[goal_id] is None and self._goal_handles[goal_id].status == GoalStatus.STATUS_EXECUTING:
                self.send_cancel_request(goal_id)
        
    def send_cancel_request(self, goal_id, cancel_timer=False):
        if not self._goal_handles[goal_id] is None and self._goal_handles[goal_id].status == GoalStatus.STATUS_EXECUTING:
            self.get_logger().info(f'Sending cancel request with argument goal_id={goal_id}')
            self._goal_handles[goal_id].cancel_goal_async() # Cancel the goal

        if cancel_timer:
            self._timer.cancel() # Cancel the timer
        
    def goal_finished(self):
        self.get_logger().info('Goal finished!')
        self._received_results += 1
        if self._sent_goals == self._received_results:
            self.get_logger().info('All goals are done!')
            rclpy.shutdown()

    def goal_response_callback(self, future, goal_id):
        self._goal_handles[goal_id] = future.result()
        if not self._goal_handles[goal_id].accepted:
            self.get_logger().warn('Goal rejected :(')
            self.goal_finished()
            return

        self.get_logger().info('Goal accepted :)')
        self._get_result_future = self._goal_handles[goal_id].get_result_async()
        self._get_result_future.add_done_callback(self.get_result_callback)

    def get_result_callback(self, future):
        result = future.result().result
        status = future.result().status

        if status == GoalStatus.STATUS_SUCCEEDED:
            self.get_logger().info('Goal succeeded!')
        elif status == GoalStatus.STATUS_CANCELED:
            self.get_logger().warn('Goal was canceled')
        elif status == GoalStatus.STATUS_ABORTED:
            self.get_logger().error('Goal was aborted')
        else:
            self.get_logger().warn('Goal failed with status: {0}'.format(status))
        
        self.get_logger().info('Result position: {0}'.format(result.position))
        
        self.goal_finished()

    def feedback_callback(self, feedback_msg):
        feedback = feedback_msg.feedback
        self.get_logger().info('Received feedback: {0}'.format(feedback.current_position))

def main(args=None):
    rclpy.init(args=args)
    action_client = RobotMoveActionClient()
    action_client.send_goal(76, 7, False)
    time.sleep(1.0)
    action_client.send_goal(0, 5)
    rclpy.spin(action_client)

if __name__ == '__main__':
    main()

File: src/actions_cpp/src/robot_move_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "my_robot_interfaces/action/robot_move.hpp"
#include "example_interfaces/msg/empty.hpp"

using RobotMove = my_robot_interfaces::action::RobotMove;
using RobotMoveGoalHandle = rclcpp_action::ClientGoalHandle<RobotMove>;
using Empty = example_interfaces::msg::Empty;
using namespace std::placeholders;

class RobotMoveClientNode : public rclcpp::Node
{
public:
    RobotMoveClientNode() : Node("move_robot_client")
    {
        move_robot_client_ = 
            rclcpp_action::create_client<RobotMove>(this, "move_robot");

        cancel_subscriber_ = this->create_subscription<Empty>(
            "cancel_move", 10, std::bind(&RobotMoveClientNode::callback_cancel_move, this, _1));
    }

    void send_goal(int position, int velocity)
    {
        move_robot_client_->wait_for_action_server();

        auto goal = RobotMove::Goal();
        goal.position = position;
        goal.velocity = velocity;

        auto options = rclcpp_action::Client<RobotMove>::SendGoalOptions();
        options.goal_response_callback = std::bind(&RobotMoveClientNode::goal_response_callback, this, _1);
        options.result_callback = std::bind(&RobotMoveClientNode::goal_result_callback, this, _1);
        options.feedback_callback = std::bind(&RobotMoveClientNode::goal_feedback_callback, this, _1, _2);

        RCLCPP_INFO(this->get_logger(), "Send goal with position %d and velocity %d", position, velocity);
        move_robot_client_->async_send_goal(goal, options);
    }

private:

    void goal_response_callback(const RobotMoveGoalHandle::SharedPtr &goal_handle)
    {
        if (!goal_handle) {
            RCLCPP_INFO(this->get_logger(), "Goal got rejected");
        }
        else {
            this->goal_handle_ = goal_handle;
            RCLCPP_INFO(this->get_logger(), "Goal got accepted");
        }
    }

    void goal_result_callback(const RobotMoveGoalHandle::WrappedResult &result)
    {
        auto status = result.code;        
        if (status == rclcpp_action::ResultCode::SUCCEEDED) {
            RCLCPP_INFO(this->get_logger(), "Succeeded");
        }
        else if (status == rclcpp_action::ResultCode::ABORTED) {
            RCLCPP_ERROR(this->get_logger(), "Aborted");
        }
        else if (status == rclcpp_action::ResultCode::CANCELED) {
            RCLCPP_WARN(this->get_logger(), "Canceled");
        }
        
        int position = result.result->position;
        std::string message = result.result->message;
        RCLCPP_INFO(this->get_logger(), "Position: %d", position);
        RCLCPP_INFO(this->get_logger(), "Message: %s", message.c_str());
    }

    void goal_feedback_callback(const RobotMoveGoalHandle::SharedPtr &goal_handle,
        const std::shared_ptr<const RobotMove::Feedback> feedback)
    {
        (void)goal_handle;
        int position = feedback->current_position;
        RCLCPP_INFO(this->get_logger(), "Feedback position: %d", position);
    }

    void callback_cancel_move(const Empty::SharedPtr msg)
    {
        (void)msg;
        cancel_goal();
    }

    void cancel_goal()
    {
        if (this->goal_handle_) {
            this->move_robot_client_->async_cancel_goal(goal_handle_);
            goal_handle_.reset();
        }
    }

    rclcpp_action::Client<RobotMove>::SharedPtr move_robot_client_;
    rclcpp::Subscription<Empty>::SharedPtr cancel_subscriber_;
    RobotMoveGoalHandle::SharedPtr goal_handle_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<RobotMoveClientNode>();
    node->send_goal(85, 6);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}